src/databricks/labs/ucx/contexts/workflow_task.py

```diff
@@ -116,10 +134,109 @@ def workflow_run_recorder(self) -> WorkflowRunRecorder:
         return WorkflowRunRecorder(
             self.sql_backend,
             self.config.ucx_catalog,
-            workspace_id=self.workspace_client.get_workspace_id(),
+            workspace_id=self.workspace_id,
             workflow_name=self.named_parameters["workflow"],
             workflow_id=int(self.named_parameters["job_id"]),
             workflow_run_id=int(self.named_parameters["parent_run_id"]),
             workflow_run_attempt=int(self.named_parameters.get("attempt", 0)),
             workflow_start_time=self.named_parameters["start_time"],
         )
+
+    @cached_property
+    def workspace_id(self) -> int:
+        return self.workspace_client.get_workspace_id()
+
+    @cached_property
+    def historical_clusters_log(self) -> HistoryLog[ClusterInfo]:
+        cluster_owner = ClusterOwnership(self.administrator_locator)
+        return HistoryLog(
+            self.sql_backend,
+            cluster_owner,
+            ClusterInfo,
+            int(self.named_parameters["parent_run_id"]),
+            self.workspace_id,
+            self.config.ucx_catalog,
+        )
```

Comment:
```suggestion
    @cached_property
    def cluster_ownership(self) -> ClusterOwnership:
        return ClusterOwnership(self.administrator_locator)
        
    @cached_property
    def historical_clusters_log(self) -> HistoryLog[ClusterInfo]:
        return HistoryLog(
            self.sql_backend,
            self.cluster_ownership,
            ClusterInfo,
            int(self.named_parameters["parent_run_id"]),
            self.workspace_id,
            self.config.ucx_catalog,
        )
```

we need `cluster_ownership` in context for https://github.com/databrickslabs/ucx/pull/2980. same for other things.

Reply:
All the ownership components have been factored out.